File: src/api/waitlist.ts
import { apiFetch } from './client'

export interface WaitlistEntry {
  id: string
  event_id: string
  event_title: string
  phase_id: string
  phase_name: string
  max_budget: string
  email: string | null
  notify_via_email: boolean
  created_at: string
}

export interface JoinWaitlistPayload {
  event_id: string
  phase_id: string
  max_budget: number
  email?: string
  notify_via_email?: boolean
}

export async function joinWaitlist(payload: JoinWaitlistPayload): Promise<WaitlistEntry> {
  return apiFetch('/api/waitlist', {
    method: 'POST',
    body: JSON.stringify(payload),
  })
}

export async function fetchMyWaitlist(): Promise<{ items: WaitlistEntry[] }> {
  return apiFetch('/api/waitlist/mine')
}

export async function leaveWaitlist(entryId: string): Promise<void> {
  return apiFetch(`/api/waitlist/${entryId}`, { method: 'DELETE' })
}

File: backend/app/schemas/waitlist.py
from datetime import datetime
from decimal import Decimal
from uuid import UUID

from pydantic import BaseModel, EmailStr, Field


class JoinWaitlistBody(BaseModel):
    event_id: UUID
    phase_id: UUID
    max_budget: Decimal = Field(..., gt=0)
    email: EmailStr | None = None
    notify_via_email: bool = True


class WaitlistEntryResponse(BaseModel):
    id: UUID
    event_id: UUID
    event_title: str
    phase_id: UUID
    phase_name: str
    max_budget: Decimal
    email: str | None
    notify_via_email: bool
    created_at: datetime

    model_config = {"from_attributes": True}


class WaitlistListResponse(BaseModel):
    items: list[WaitlistEntryResponse]
